### PR TITLE
Adding installation support (again)

### PR DESCRIPTION
**Pull request by sean-parent**
Added cleaned up installation support, along with an overhauled readme.

---

> [!NOTE]
> Adds installation and CMake package-config generation (with dependency mapping/merging), new CI to verify install/find_package, comprehensive README and setup script, and unit tests for dependency handling.
> 
> - **Install & Packaging**:
>   - New `cmake/cpp-library-install.cmake` with `cpp_library_map_dependency()` and automatic `find_dependency()` generation (version detection, system-package handling, component merging) for installed `Config.cmake`.
>   - `templates/Config.cmake.in` updated to include generated `@PACKAGE_DEPENDENCIES@` and correct `Targets.cmake` include.
> - **Core CMake**:
>   - `cpp-library.cmake`/`cpp-library-setup.cmake`: integrate install when top-level; derive `PACKAGE_NAME`; support `CPP_LIBRARY_VERSION` override; shared/static via `BUILD_SHARED_LIBS`.
>   - Template copying now generates CI from `templates/.github/workflows/ci.yml.in` via new `cmake/cpp-library-ci.cmake`.
> - **CI/CD**:
>   - Add repo CI `.github/workflows/ci.yml`: unit tests (CMake script), integration test that builds/installs a sample lib and consumes it via `find_package`, and docs checks.
>   - Template CI revamped (`templates/.github/workflows/ci.yml.in`) and presets add `install` configuration.
> - **Docs & Tooling**:
>   - Major README rewrite: quick start, manual setup, installation/consumption, dependency mapping, versioning, troubleshooting.
>   - New `setup.cmake` interactive bootstrap script.
>   - Minor template tweaks (`templates/.vscode/extensions.json`).
> - **Tests**:
>   - Add `tests/install/` with 18 unit tests covering dependency mapping, component merging, system packages, and edge cases, plus summary docs.
> - **Misc**:
>   - `.gitignore` adds `.DS_Store`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 256d0e7a5472ee1dfb07198175278eff2c9daf95. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: templates/.github/workflows/ci.yml [hunk 1]
```diff
@@ -43,6 +43,51 @@ jobs:
...
+          # Create test CMakeLists.txt
+          cat > CMakeLists.txt << EOF
+          cmake_minimum_required(VERSION 3.20)
+          project(test-find-package CXX)
+
+          set(CMAKE_PREFIX_PATH "${INSTALL_PREFIX}")
+          find_package(${PACKAGE_NAME} REQUIRED)
+
+          message(STATUS "Successfully found ${PACKAGE_NAME}")
+          EOF
```
[the remark this answers is not in the document]
Comment (sean-parent) [reply templates/.github/workflows/ci.yml]: This is not an issue - the leading text is stripped by the YAML parser.

File: cmake/cpp-library-install.cmake [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        NAMESPACE   # Namespace for alias (e.g., "stlab")
+    )
+    set(multiValueArgs
+        HEADERS     # List of header file paths (for FILE_SET support check)
+    )
+    
+    cmake_parse_arguments(ARG "" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+    
+    # Validate required arguments
+    if(NOT ARG_NAME)
```
Comment (Ikar-Rahl) [inline cmake/cpp-library-install.cmake]: ARG_NAME is setup higher as set(ARG_NAME "${PROJECT_NAME}").
For copy-on-write, this would be stlab-copy-on-write, which generate a stlab-copy-on-writeConfig.cmake, targets, version etc.
This is also generating add_library(stlab::stlab-copy-on-write INTERFACE IMPORTED) in the targets;
I guess this is unwanted, CPM make a call for find_package(copy-on-write), not stlab-copy-on-write.
Same for stlab, it links against stlab::copy-on-write, not stlab::stlab-copy-on-write.

You can either:
- Change the project name for copy-on-write etc.
- remove the stlab when doing the set(ARG_NAME "${PROJECT_NAME}")
- add a project_name to cpp_library_setup

**Review by Ikar-Rahl — COMMENTED** [review]
Thanks a lot for your work ! I am happy to participate if needed. :-)

tested on my vcpkg environment, got stuck on stlab, but otherwise promising !

File: cmake/cpp-library-install.cmake [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    # Uses SameMajorVersion compatibility (e.g., 2.1.0 is compatible with 2.0.0)
+    write_basic_package_version_file(
+        "${CMAKE_CURRENT_BINARY_DIR}/${ARG_NAME}ConfigVersion.cmake"
+        VERSION ${ARG_VERSION}
+        COMPATIBILITY SameMajorVersion
+    )
+    
+    # Generate package config file from template
+    configure_file(
+        "${CPP_LIBRARY_ROOT}/templates/Config.cmake.in"
```
[the remark this answers is not in the document]
Comment (Ikar-Rahl) [reply cmake/cpp-library-install.cmake]: I though about just using CPM_PACKAGES and CPM_PACKAGE_*_VERSION for theses, but there is one main issue with that, CPM does not indicate anywhere if a dependency is public or private cpp_library and enum are good 2 examples, one is for internal use only, another has to be exposed, and CPM does not have this information.
We could, of course, make workarounds, keeping a hard coded list of private deps like cpp_library. But I feel it a bit shaky.

There is also benefit that all you dependency go through the same function call, no matter the flavor.

It is up to you to decide what you prefer, I can implement it either way.

File: cmake/cpp-library-install.cmake (hunk 2 above)
Comment (sean-parent) [reply cmake/cpp-library-install.cmake]: Good catch - pushed a fix for this. (will test before I commit this PR).

File: cmake/cpp-library-install.cmake (hunk 3 above)
[the remark this answers is not in the document]
Comment (sean-parent) [reply cmake/cpp-library-install.cmake]: Okay - I updated the auto dependency() call generation to allow a map for custom dependency formats. This way, I'm not on the hook to keep a custom table up-to-date.

File: cmake/cpp-library-setup.cmake [hunk 4]
```diff
@@ -2,7 +2,8 @@
 #
 # cpp-library-setup.cmake - Core library setup functionality
 
-# Function to get version from git tags
+# Returns version string from latest git tag, falling back to "0.0.0".
+# - Postcondition: OUTPUT_VAR set to version string with 'v' prefix removed
 function(_cpp_library_get_git_version OUTPUT_VAR)
     # Try to get version from git tags
     execute_process(
```
[the remark this answers is not in the document]
Comment (Ikar-Rahl) [reply cmake/cpp-library-setup.cmake]: Thanks, work perfectly :-)

**Review by Ikar-Rahl — CHANGES_REQUESTED** [review]
So, with the following changes:
- ${PKG_NAME}-${COMPONENT} => ${COMPONENT} 
- project(enum-ops) to project(stlab-enum-ops)
- CPMAddPackage("gh:stlab/copy-on-write@1.0.6") => CPMAddPackage("gh:stlab/stlab-copy-on-write@1.0.6"), same for enum.
- Used _cpp_library_setup_install in the stlab, with minor renaming of FILE_SET to headers.
I was able to make it work full circle.

stlab produced a correclty formed Config package:

> # Find dependencies required by this package
> find_dependency(stlab-copy-on-write)
> find_dependency(stlab-enum-ops)
> find_dependency(Threads)
> find_dependency(Qt6)

Final decision about naming still to be made, but this is viable.
